includes/controllers/class-llms-controller-checkout.php

```diff
@@ -0,0 +1,404 @@
+		}
+
+		$this->send_json( $handle );
+
+	}
+
+	/**
+	 * Extracts data from `$_POST` into an array that can be passed into `llms_setup_pending_order()`.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $posted_data Data array, from `$_POST`.
+	 * @return array
+	 */
+	private function extract_setup_data( $posted_data ) {
+
+		$plan_id = absint( $posted_data['llms_plan_id'] ?? 0 );
+
+		$data = array(
+			'plan_id'         => $plan_id,
+			'agree_to_terms'  => llms_parse_bool( $posted_data['llms_agree_to_terms'] ?? '' ),
+			'coupon_code'     => sanitize_text_field( $posted_data['llms_coupon_code'] ?? '' ),
+			'customer'        => $this->extract_user_data( $posted_data, $plan_id ),
+			'payment_gateway' => sanitize_text_field( $posted_data['llms_payment_gateway'] ?? '' ),
+		);
+
+		return $data;
+
+	}
+
+	/**
+	 * Extracts user registration / update information from a posted data array.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $posted_data Raw $_POST (or similar) data.
+	 * @return array
+	 */
+	private function extract_user_data( $posted_data, $plan_id ) {
+
+		$user_data = array();
+		$plan      = $plan_id ? llms_get_post( $plan_id ) : false;
+
+		$uid = get_current_user_id();
+		if ( $uid ) {
+			$user_data['user_id'] = $uid;
+		}
+
+		foreach ( LLMS_Forms::instance()->get_form_fields( 'checkout', compact( 'plan' ) ) as $field ) {
+			if ( isset( $posted_data[ $field['name'] ] ) ) {
+				$user_data[ $field['name'] ] = $posted_data[ $field['name'] ];
+			}
+		}
+
+		return $user_data;
+
+	}
+
+	/**
+	 * Retrieves the AJAX url for the requested action
+	 *
+	 * @since [version]
+	 *
+	 * @param string $action A checkout action. Accepts class constant `ACTION_CREATE_PENDING_ORDER` or `ACTION_CONFIRM_PENDING_ORDER`.
+	 * @return string
+	 */
+	public function get_url( $action ) {
+		return add_query_arg(
+			self::AJAX_QS_VAR,
+			wp_create_nonce( $action ),
+			get_site_url()
+		);
+	}
+
+	/**
+	 * Sends a JSON response.
+	 *
+	 * @since [version]
+	 *
+	 * @param array|WP_Error $data Response data.
+	 * @return void
+	 */
+	private function send_json( $data ) {
+
+		// Tell WP we're doing AJAX.
+		add_filter( 'wp_doing_ajax', '__return_true' );
+		wp_send_json( $data, is_wp_error( $data ) ? 400 : 200 );
+
+	}
+
+	/**
+	 * Verifies an incoming request nonce and posted action field.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $field The nonce field.
+	 * @param string $nonce The nonce & action value.
+	 * @return null|bool Returns `null` if the nonce isn't submitted or can't be verified, `false` if the
+	 *                   action isn't submitted or doesn't match the intended action, and `true` if
+	 *                   the request is verified successfully.
+	 */
+	private function verify_request( $field, $nonce ) {
+
+		if ( ! llms_verify_nonce( $field, $nonce, 'POST' ) ) {
+			return null;
+		}
+
+		if ( llms_filter_input( INPUT_POST, 'action' ) !== $nonce ) {
+			return false;
+		}
```

Comment:
`llms_verify_nonce()` not only verifies the nonce but also verifies that the form has been submitted. Frankly it doesn't really matter which one runs first. Historically with form submission controllers I've followed a pattern (yes, it's not documented) where `null` is returned if the form wasn't submitted (in my mind this has always meant if the request method doesn't match and the nonce isn't found in the request payload). Returning `null` because the form wasn't submitted and also if the nonce is invalid.

The semantic difference between `null` and `false` really doesn't matter. I have different returns so I can easily differentiate the different conditions in testing. We could make the returns `1`, `2,` and `3` if we wanted. Or we could return `null` (or `false` for both conditions using an `||` operator but I wanted to be able to test each case independently.

Reply:
yessss!